[RFC FS-1071] Witnesses passing for trait-constraints w.r.t. quotations

dsyme opened the pull request:
Implementation for [RFC FS-1071 - Witness passing for quotations using SRTP-constrained functions and methods](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1071-witness-passing-quotations.md)

F# quotations of code using SRTP constraint trait calls don't contain sufficient information to represent the semantic intent of the code. Specifically the quotations are missing any record of the resolution of SRTP constraints or how to execute constraint calls. 

This RFC addresses this problem by incorporating the necessary information into both quotations and the code laid down for dynamic interpretation of quotations.

Continuation of https://github.com/dotnet/fsharp/pull/6345

cartermp commented:
@dsyme given the outstanding TODOs should this be marked as WIP?

dsyme commented:
> @dsyme given the outstanding TODOs should this be marked as WIP?

Yes.  The core of the work is cracked but there are some painful edge cases (super-rare, but they should still be dealt with)

dsyme commented:
Current status: failing test
```
         matchbangquot1 OK
          check CallWithWitness OK
          check CallWithWitness (DateTime + TimeSpan) OK
          check Call (DateTime + TimeSpan) OK
          System.InvalidOperationException: Could not bind to method
             at Microsoft.FSharp.Quotations.PatternsModule.bindMethodBySearch(Type parentT, String nm, Int32 marity, FSharpList`1 argtys, FSharpFunc`2 rty) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 1058
             at Microsoft.FSharp.Quotations.PatternsModule.bindMeth(Tuple`5 _arg1, FSharpList`1 tyargs) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 1281
             at Microsoft.FSharp.Quotations.PatternsModule.u_opSpec@1705-27.Invoke(FSharpList`1 tyargs) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 1707
             at Microsoft.FSharp.Quotations.PatternsModule.u_Expr@1508-4.Invoke(BindingEnv env) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 1517
             at Microsoft.FSharp.Quotations.PatternsModule.u_Expr@1530-9.Invoke(BindingEnv env) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 1530
             at Microsoft.FSharp.Quotations.PatternsModule.deserialize(Type localAssembly, Type[] referencedTypeDefs, Type[] spliceTypes, FSharpExpr[] spliceExprs, Byte[] bytes) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 1913
             at <StartupCode$FSI_0003>.$FSI_0003.main@()
          Stopped due to error
```

dsyme commented:
One failure here:
```
          D:\a\1\s\.dotnet\dotnet.exe D:\a\1\s\artifacts\bin\fsi\release\netcoreapp2.1\fsi.exe --define:NETCOREAPP
          --define:RELEASE
          --define:NETCOREAPP
          --define:NETCOREAPP2_0
          --optimize+
          --warn:3
          --warnaserror:3239,76
          --fullpaths
          --flaterrors
          --targetprofile:netcore
          D:\a\1\s\tests\fsharp\coreclr_utilities.fs
          D:\a\1\s\tests\fsharp\tools\eval\test.fsx

          System.TypeInitializationException: The type initializer for 'BinaryOpDynamicImplTable`4' threw an exception. ---> System.NullReferenceException: Object reference not set to an instance of an object.
             at Microsoft.FSharp.Core.LanguagePrimitives.BinaryOpDynamicImplTable`4..cctor() in D:\a\1\s\src\fsharp\FSharp.Core\prim-types.fs:line 3194
             --- End of inner exception stack trace ---
             at Microsoft.FSharp.Core.LanguagePrimitives.AdditionDynamic[T1,T2,TResult](T1 x, T2 y) in D:\a\1\s\src\fsharp\FSharp.Core\prim-types.fs:line 3233
             at <StartupCode$FSI_0002>.$FSI_0002.main@()
          Stopped due to error
```

dsyme commented:
Failures: 
```
          wekncjeck2 OK
          wekncjeck3 FAILED: got "Call (Some (Lambda (arg0_0,
                              Lambda (arg1_0, Call (None, op_Addition, [arg0_0, arg1_0])))),
                F, [Value (3)])", expected "Call (Some (NewObject (C)), F, [Value (3)])" NO: wekncjeck3
          wekncjeck4 OK
          wekncjeck5 OK
          wekncjeck6 OK
          wekncjeck7 OK
          ewlknweknl1 OK
          ewlknweknl2 OK
          System.ArgumentException: Type mismatch when building 'args': invalid parameter for a method or indexer property. Expected 'Microsoft.FSharp.Core.FSharpFunc`2[System.Int32,Microsoft.FSharp.Core.FSharpFunc`2[System.Int32,System.Int32]]', but received type 'FSI_0003+MoreWitnessTests+Tests+C'.
          Parameter name: receivedType
             at Microsoft.FSharp.Quotations.PatternsModule.checkTypesWeakSR[a](Type expectedType, Type receivedType, a name, String threeHoleSR) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 734
             at Microsoft.FSharp.Quotations.PatternsModule.checkArgs@739.Invoke(ParameterInfo p, FSharpExpr a) in D:\a\1\s\src\fsharp\FSharp.Core\quotations.fs:line 739
             at Microsoft.FSharp.Collections.ListModule.loop@191-27[T1,T2](FSharpFunc`3 f, FSharpList`1 list1, FSharpList`1 list2) in D:\a\1\s\src\fsharp\FSharp.Core\list.fs:line 194


Failed   eval-FSC_BASIC

Failed   queriesOverIQueryable

test.fsx(1908,49): error FS0450: Quotations cannot contain expressions that take the address of a field

test.fsx(1926,54): error FS0450: Quotations cannot contain expressions that take the address of a field

test.fsx(1933,40): error FS0450: Quotations cannot contain expressions that take the address of a field

Failed   queriesLeafExpressionConvert

System.NotSupportedException: Could not convert the following F# Quotation to a LINQ Expression Tree
--------
AddressOf (copyOfStruct)
-------------

   at Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in D:\a\1\s\src\fsharp\FSharp.Core\Linq.fs:line 823
   at Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvObjArg(ConvEnv env, FSharpOption`1 objOpt, FSharpOption`1 coerceTo) in D:\a\1\s\src\fsharp\FSharp.Core\Linq.fs:line 838
   at Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in D:\a\1\s\src\fsharp\FSharp.Core\Linq.fs:line 485
   at Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.ConvExprToLinqInContext(ConvEnv env, FSharpExpr inp) in D:\a\1\s\src\fsharp\FSharp.Core\Linq.fs:line 801
   at Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter.QuotationToExpression(FSharpExpr e) in D:\a\1\s\src\fsharp\FSharp.Core\Linq.fs:line 854
   at FSI_0001.LeafExpressionEvaluationTests.Eval[T](FSharpExpr`1 q) in D:\a\1\s\tests\fsharp\core\queriesLeafExpressionConvert\test.fsx:line 45
   at FSI_0001.LeafExpressionEvaluationTests.checkEval[T](String nm, FSharpExpr`1 q, T expected) in D:\a\1\s\tests\fsharp\core\queriesLeafExpressionConvert\test.fsx:line 51
   at <StartupCode$FSI_0001>.$FSI_0001.main@() in D:\a\1\s\tests\fsharp\core\queriesLeafExpressionConvert\test.fsx:line 167

```

cartermp commented:
The diff in prim-types is enormous. Is that intended?

dsyme commented:
@KevinRansom I assume you want this targeted at `release/fsharp5`?

dsyme commented:
@KevinRansom I assume you want this targeted at `release/fsharp5`?

dsyme commented:
@KevinRansom @brettfo I'm seeing this error on branches like this one

> ##[error]src\fsharp\fsi\fsi.fsproj(0,0): error MSB4057: (NETCORE_ENGINEERING_TELEMETRY=Build) The target "UpdateXlf" does not exist in the project.

It seems to come up even after entirely harmless integration commits like https://github.com/dotnet/fsharp/pull/6810/commits/ae0e88b8b00be1b9b28ede683c84c710317f55ef.  See how that commit turned from green to red:

![image](https://user-images.githubusercontent.com/7204669/70992622-fdf9b980-20c1-11ea-8698-d75589010366.png)

Any ideas??

KevinRansom commented:
That is a heisenbug we haven’t been able to track down.  It’s something to do with arcade we believe.

From: Don Syme <notifications@github.com>
Sent: Tuesday, December 17, 2019 3:41 AM
To: dotnet/fsharp <fsharp@noreply.github.com>
Cc: Kevin Ransom <Kevin.Ransom@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [dotnet/fsharp] [RFC FS-1071] Witnesses passing for trait-constraints w.r.t. quotations (#6810)


@KevinRansom<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FKevinRansom&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C598e26cd33bb41b01ae308d782e5f003%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637121796394692968&sdata=9vSmhTgpVTY2OEZupNYU%2FQL%2B72vBpN0i7wJ2z%2BONLfc%3D&reserved=0> @brettfo<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbrettfo&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C598e26cd33bb41b01ae308d782e5f003%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637121796394702973&sdata=QbFzgrWZGI%2BGJx%2FeRuDgnzSx97zBJHvUtcGIAdggaZQ%3D&reserved=0> I'm seeing this error on branches like this one

##[error]src\fsharp\fsi\fsi.fsproj(0,0): error MSB4057: (NETCORE_ENGINEERING_TELEMETRY=Build) The target "UpdateXlf" does not exist in the project.

It seems to come up even after entirely harmless integration commits like ae0e88b<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Ffsharp%2Fcommit%2Fae0e88b8b00be1b9b28ede683c84c710317f55ef&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C598e26cd33bb41b01ae308d782e5f003%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637121796394702973&sdata=cELkpkgAIv16jYwGUwFpMkch07vS47mSdvUjpVZwUuY%3D&reserved=0>. See how that commit turned from green to red:

[image]<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F7204669%2F70992622-fdf9b980-20c1-11ea-8698-d75589010366.png&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C598e26cd33bb41b01ae308d782e5f003%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637121796394712964&sdata=7u8gkUgknHM0PMNAmD%2BGVq5bLMzq8h%2BQlheU2%2FQnJO8%3D&reserved=0>

Any ideas??

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Ffsharp%2Fpull%2F6810%3Femail_source%3Dnotifications%26email_token%3DABHPUFVCAQK2E4NQ3FGJVLLQZC3DLA5CNFSM4HOTF5AKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHCDEEY%23issuecomment-566506003&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C598e26cd33bb41b01ae308d782e5f003%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637121796394712964&sdata=37GpJiNtBGjImViIonFodOz97n6G%2FNo9hAGLzraCN3Y%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABHPUFSFMI3EZ3EI6PJESLLQZC3DLANCNFSM4HOTF5AA&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C598e26cd33bb41b01ae308d782e5f003%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637121796394722960&sdata=m3kQ%2ByYbtTQGVAQTk75f6B6%2FyBKhoe9U573oB7mBnPE%3D&reserved=0>.

dsyme commented:
The only place I see anything to do with UpdateXlf is here:
```
  <PropertyGroup>
    <EnableXlfLocalization Condition="'$(EnableXlfLocalization)' == '' AND ('$(Configuration)' == 'Proto' OR '$(MonoPackaging)' == 'true')">false</EnableXlfLocalization>
    <UpdateXlfOnBuild Condition="'$(CI)' != '1'">true</UpdateXlfOnBuild>
  </PropertyGroup>
```

dsyme commented:
> The diff in prim-types is enormous. Is that intended?

@cartermp We need to discuss this.  I'm torn about a few things in this RFC and its implementation but in my heart I know that it is logically the right thing and (roughly speaking) the right way to solve this problem at the core.

I'll try to capture the issues in the RFC and then lets discuss.  Basically to make interpreting quotations smooth and reliable we need some kind of executable witnesses for all the "built in" ops we've traditionally compiled away through inlining code down to IL ops.  There are many more of these than might be appreciated, e.g. left shift operations on all the integer sizes, etc. etc. 

By an executable witness I mean a method we can call at runtime to perform a primitive operation corresponding the satisfaction of a primitive constraint.  For example, if the "left shift" operator is called in unsigned 16 bit integers we need a method we can call to actually do that left shift. If the UInt16 type actually had the left shift operation defined on it then it would be easy.  But it doesn't.

There are (to some extent) executable witnesses available for (some of) these in the .NET core libraries but there is nothing systematic. 

It's a really nasty problem - but quotations will never be a fully reliable basis for meta-programming  in the presence of SRTP operations like `+` and `-` (including the use of quotations for compile-time meta-programming when writing type providers) until we get a grip on this fundamental issue.

It could be we could continue to rely on adhoc resolution in clients of quotation processors for "built in" cases and only use the witness passing for the more nasty unresolvable cases where only the F~ compiler knows the answer.  But considerthis code in a quotation consumer:

https://github.com/dotnet/fsharp/pull/6810/files#diff-5ce60fdfe5066ce67d7dd5d354792434R514-R548

or this

https://github.com/fsprojects/FSharp.TypeProviders.SDK/blob/master/src/ProvidedTypes.fs#L14153-L14737

where 600+ lines are required to list out the different cases _in every consumer of quotations_, when in contrast you could instead just process a single case of `CallWithWitnesses` where the quotation contains the answer and a method to call:

https://github.com/dotnet/fsharp/pull/6810/files#diff-5ce60fdfe5066ce67d7dd5d354792434R650

dsyme commented:
> That is a heisenbug we haven’t been able to track down.  It’s something to do with arcade we believe.

Cool, thanks, understood.

I've seen one other heisenbug, will try to record it here too

dsyme commented:
@cartermp I've simplified the change to prim-types so it doesn't add anywhere near as many new public methods.

dsyme commented:
I've now resolved all the open outstanding design issues for this PR, updated the testing and updated the RFC.

It's a significant chunk of work and I've been working to minimize the diff.  I can probably minimize it a bit further (there are some useful cosmetic changes in IlxGen.fs and Linq.fs which I could minimize further).

I'd appreciate an initial cosmetic review at some point to help guide me to make another iteration.

dsyme commented:
I've also had a chance to talk this whole RFC over with @eiriktsarpalis, and he confirms it would have been useful for any of his prior working with quotations.

dsyme commented:
Yay, this one is green at last.  I've added the "Candidate for F# 5" label

cartermp commented:
@dsyme note that we're using that label for work that is no longer WIP - ready for review with aim to merge and release in a preview. The `vNext` label might be more appropriate at this stage.

cartermp commented:
Ah I see you've removed the WIP

dsyme commented:
> @dsyme note that we're using that label for work that is no longer WIP - ready for review with aim to merge and release in a preview. The vNext label might be more appropriate at this stage.

Yes, this is ready for review.  I'd expect an iteration or two but I've nothing more specific to do on it

cartermp commented on src/fsharp/MethodCalls.fs:
```diff
@@ -1826,19 +1826,20 @@ let GenWitnessExpr amap g m (traitInfo: TraitConstraintInfo) argExprs =
             // fix - apply coercion for the arguments (excluding 'receiver' argument in instance calls)
 
-            // flatten list of argument types (looks like trait calls with curried arguments are not supported so we can just convert argument list in straightforward way)
+            // flatten list of argument types (looks like trait calls with curried arguments are not supported so we can just convert argument list in straighforward way)
```
```suggestion
            // flatten list of argument types (looks like trait calls with curried arguments are not supported so we can just convert argument list in straightforward way)
```

cartermp commented on src/fsharp/MethodCalls.fs:
```diff
@@ -1826,19 +1826,20 @@ let GenWitnessExpr amap g m (traitInfo: TraitConstraintInfo) argExprs =
         let argExprs = 
             // FIX for #421894 - typechecker assumes that coercion can be applied for the trait calls arguments but codegen doesn't emit coercion operations
-            // result - generation of non-verifiable code
+            // result - generation of non-verifyable code
```
```suggestion
            // result - generation of non-verifiable code
```

dsyme replied:
Thanks

cartermp commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -380,9 +385,21 @@ namespace Microsoft.FSharp.Core
+
+        type System.Type with
+            
+            // Note, this differs slightly from the nettstandard1.6 FX_RESHAPED_REFLECTION case as it looks for non-public methods.
```
Not sure if this helps, but all that stuff is now removed since FSharp.Core is NS2.0 only now.

dsyme replied:
Thanks, I've removed the comment

cartermp commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -2336,7 +2357,22 @@ namespace Microsoft.FSharp.Core
+
+        let inline ParseIntPtr (s:string)  = (# "conv.ovf.i"  (ParseInt64 s)  : nativeint #)
+
+        let inline ParseUIntPtr (s:string) = (# "conv.ovf.u"  (ParseInt64 s)  : unativeint #)
```
Should these just use the .NET methods?

dsyme replied:
For the purposes of this PR let's ignore this, as the code is just moved around to put the dynamic invocation tables together

cartermp commented on src/fsharp/FSharp.Core/quotations.fs:
```diff
@@ -1456,13 +1543,39 @@ module Patterns =
+        let meths = ty.GetMethods staticBindingFlags |> Array.filter (fun mi -> mi.Name = nm)
+        match meths with
+        | [||] ->
+            raise <| System.InvalidOperationException (String.Format(SR.GetString(SR.QcannotBindFunction), nm, ty.ToString()))
```
`invalidOp`

cartermp commented on src/fsharp/FSharp.Core/quotations.fs:
```diff
@@ -1456,13 +1543,39 @@ module Patterns =
+                let methsW = ty.GetMethods(staticBindingFlags) |> Array.filter (fun mi -> mi.Name = nmW)
+                match methsW with
+                | [||] ->
+                    raise <| System.InvalidOperationException (String.Format(SR.GetString(SR.QcannotBindFunction), nmW, ty.ToString()))
```
`invalidOp`

dsyme replied:
ok thanks

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -834,7 +841,10 @@ and IlxGenEnv =
 
-      /// For optimizing direct tail recursion to a loop - mark says where to branch to. Length is 0 or 1.
+      /// All witnesses in scope
+      witnessesInScope: ImmutableDictionary<TraitWitnessInfo, ValStorage>
```
While I'm ok with using `ImmutableDictionary`, we tend to be using F#'s `Map`.

What was the reason for choosing `ImmutableDictionary` instead of `Map` here? Performance?

dsyme replied:
It is because the dictionary gets an IEqualityComparer that captures `TcGlobals`, here
```fsharp
      witnessesInScope= 
          ImmutableDictionary.Create(
               { new IEqualityComparer<_> with 
                      member __.Equals(a, b) = traitKeysAEquiv g TypeEquivEnv.Empty a b
                      member __.GetHashCode(a) = hash a.MemberName
               })
```

If the `TraitWitnessInfo` objects captured `TcGlobals` themselves and could thus implement their own internalized equality/hash logic. However even then we couldn't use a `Map` becuase they are built on sorted keys using IComparer, rather than hashing using IEqualityComparer.

I'll add a note to the source

dsyme commented on src/fsharp/FSharp.Core/prim-types.fsi:
```diff
@@ -2747,6 +2825,7 @@ namespace Microsoft.FSharp.Core
         /// input types the operation requires an appropriate static conversion method on the input type.</summary>
         /// <param name="value">The input value.</param>
         /// <returns>The converted char.</returns>
+        [<NoDynamicInvocation(isLegacy=true)>]
```
The implementation always had this and it should have been present in the signature as well (at least, for consistency with the other cases).  It doesn't affect generated code

dsyme commented:
> Quick pass - there's a lot of inline IL added here. Is it possible to have that just use well-known .NET methods instead? Someone who hasn't memorized IL instructions can't review it easily

This is hard.  I guess you're referring to all of this https://github.com/dotnet/fsharp/pull/6810/files#diff-5470ddaa52d99663f8d391a77ae04f0dR2532

Each case is taken precisely from the corresponding primitive implementation such as here: https://github.com/dotnet/fsharp/pull/6810/files#diff-5470ddaa52d99663f8d391a77ae04f0dR4357.  These primitive implementations are *the* semantics of the operations and these are the exact things we need to lift to witnesses.

I could have pulled each of these little sequence out into a little inline method to avoid any duplication but I decided it's just not worth it - there is just as much risk of making transcription errors and the sequences are nearly always just one opcode, plus I don't really want to chrun the existing implementations, nor add the extra (internal) method overhead (very small but still...)

We do have good coverage of all these cases particularly through the `eval` quotation tests that evaluate quotations of all basic operators at all types using the witness-passing implementations.

TIHan commented on src/fsharp/MethodCalls.fs:
```diff
@@ -1895,8 +1898,36 @@ let GenWitnessExpr amap g m (traitInfo: TraitConstraintInfo) argExprs =
+        Choice2Of2 (mkMemberLambdas m [] None None vsl (expr, tyOfExpr g expr))
+    | None -> 
+        Choice1Of2 witnessInfo
+        //assert ("A constraint witness could not be found for a built-in constraint solution" |> ignore; false)
```
nit: Two lines of commented code should be removed.

dsyme replied:
thanks!

dsyme commented:
>All of the FSharp.Core API additions will need protection with this:

Thanks @cartermp,  now done

TIHan commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -316,10 +316,15 @@ namespace Microsoft.FSharp.Core
-    [<AttributeUsage(AttributeTargets.Parameter, AllowMultiple=false)>]
+        new () = NoDynamicInvocationAttribute(false)
+
+        member x.IsLegacy = isLegacy
```
From our conversation, we think we may not need "IsLegacy". When set to "true", it would tell the compiler to actually generate the implementation of the method but keep the signature with the attreibute; but, we might be able to remove "[<NoDynamicInvocation>]" instead of having to add "IsLegacy".

dsyme replied:
See below

TIHan commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -3366,9 +3866,9 @@ namespace Microsoft.FSharp.Core
         let defaultValueArg arg defaultValue = match arg with ValueNone -> defaultValue | ValueSome v -> v
 
-        [<NoDynamicInvocation>]
+        [<NoDynamicInvocation(isLegacy=true)>]
```
Were we not able to get rid of "[<NoDynamicInvocation>]" for these functions?

dsyme replied:
Working on it

dsyme replied:
@TIHan I looked at this again.  

1. The attribute is still needed to codegen the exception-raising stub body for some non-legacy cases such as the functions in `NatrivePtr.fs`

2. The attribute is needed with Legacy=true for generating the exception-raising stub body for the non-witness-passing entry points for the other cases

The generating code is here:
```
                let attr = TryFindFSharpBoolAttributeAssumeFalse cenv.g cenv.g.attrib_NoDynamicInvocationAttribute v.Attribs
                if (not hasWitnessArgs && attr.IsSome) ||
                   (hasWitnessArgs && attr = Some false) then
                    let exnArg = mkString cenv.g m (FSComp.SR.ilDynamicInvocationNotSupported(v.CompiledName g.CompilerGlobalState))
                    let exnExpr = MakeNotSupportedExnExpr cenv eenv (exnArg, m)
                    mkThrow m returnTy exnExpr
                else
                    body
```

I've removed the attribute from the `prim-types.fsi` and `nativeptr.fsi` signature files since it is not relevant "from the outside", and only relevant to guiding codegen for FSharp.Core.

I think that resolves this point.

TIHan commented:
I'm still going through everything. A debug assert got hit and trying to figure out what to do.

TIHan commented on tests/fsharp/core/quotes/test.fsx:
```diff
@@ -3161,6 +3161,681 @@ module TestMatchBang =
+    open System.Runtime.CompilerServices
+    open System.IO
+
+    // TODO - ths fails
```
Is it ok that this would be failing?

dsyme commented:
@TIHan I've fixed the case above - thank you.

Although it's simple we didn't have test cases - there are a lot of cases for inline code causing witness passing, and lots for quotation literals `<@ ... @>`, but not for inline code that contains quotations that capture witness passing.

The fix needs a bit more attention tomorrow, but I wanted you to know it's basically fixed and there's some progress here.

dsyme commented:
> Another thing, since we are keeping isLegacy, is there intention for library authors to make use of it? If not, it is still public and authors can use it which would make me a little worried.

Thanks, I made it internal

dsyme commented:
OK, this is ready for next review (Assuming it goes green, CI seems very slow at the moment)

TIHan commented:
Ok, I should be done with the review again soon.

TIHan commented on src/fsharp/FSharp.Core/prim-types.fsi:
```diff
@@ -715,7 +715,7 @@ namespace Microsoft.FSharp.Core
         new : unit -> NoDynamicInvocationAttribute
 
         [<Experimental("Experimental library feature, requires '--langversion:preview'")>]
```
Because this is internal, we don't need the `Experimental` attribute.

dsyme replied:
done thanks

TIHan commented on src/fsharp/TypedTreeOps.fs:
```diff
@@ -5695,6 +5774,32 @@ let mkArrayType (g: TcGlobals) ty = TType_app (g.array_tcr_nice, [ty])
+    let argtys = if argtys.IsEmpty then [g.unit_ty] else argtys
+    let argtysl = List.map List.singleton argtys
+    argtysl
+    //match tys with 
```
This commented code should probably deleted.

dsyme replied:
done thanks

TIHan commented on tests/fsharp/tools/eval/test.fsx:
```diff
@@ -1477,6 +1477,269 @@ module EvaluationTests =
+
+        test5 3s 4s
+        test5 3L 4L
+(*
```
Was this meant to be commented out?

dsyme replied:
No,thanks - I've uncommented it (will see in CI if it causes problems)

dsyme commented:
@TIHan I've pushed a fix and better testing for the abs bug and addressed the other problems

thanks
don

dsyme commented:
We'll need to check CI again.  There was an unrelated failure (something to do with F# Interactive).  Also the changes need careful checking.

dsyme commented:
OK, all green - @TIHan Can you review and then we should be good to go!

TIHan commented:
Looking at this now. Will be testing the latest changes.

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -960,16 +979,28 @@ let StorageForVal g m v eenv =
+    match eenv.witnessesInScope.TryGetValue w with 
+    | true, storage -> Some storage
+    | _ ->
+        //let generateWitnesses = ComputeGenerateWitnesses g eenv
```
nit: Remove commented code.

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -3055,7 +3111,61 @@ and GenUntupledArgExpr cenv cgbuf eenv m argInfos expr sequel =
+        match witnessExpr with
+        | Choice1Of2 _traitInfo ->
+            System.Diagnostics.Debug.Assert(false, "expected storage for witness")
+            //failwith "unexpected non-generation of witness "
```
nit: Remove commented code. 

Also, since this should not happen, shouldn't we actually fail?

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -3055,7 +3111,61 @@ and GenUntupledArgExpr cenv cgbuf eenv m argInfos expr sequel =
+    match storage with 
+    | None ->
+        System.Diagnostics.Debug.Assert(false, "expected storage for witness")
+        //failwith "unexpected non-generation of witness "
```
nit: Remove commented code.

Also, since this should not happen, shouldn't we actually fail?

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -3066,34 +3176,43 @@ and GenApp cenv cgbuf eenv (f, fty, tyargs, curriedArgs, m) sequel =
 
             let numArgs = List.sum arityInfo
 
+            // TODO: witness argument generation for closures
```
Will doing the work for this "TODO" later be another language feature and/or break compat?

TIHan replied:
Just asking, as I'm having a bit of a time understanding this whole bit in "GenApp".

dsyme replied:
This TODO is not real - I have removed it.  Closures capture sitnesses but do not take them as arguments at their invocation methods

I think we should clean up GenApp and separate its branches

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -4579,6 +4740,33 @@ and GetIlxClosureFreeVars cenv m (thisVars: ValRef list) eenvouter takenNames ex
+    let cloWitnessInfos = 
+        let generateWitnesses = ComputeGenerateWitnesses g eenvinner
+        if generateWitnesses then 
+            GetTraitWitnessInfosOfTypars g 0 cloFreeTyvars // TODO: 0 may be wrong here
```
I know that, "0", represents the number of parent typars and we are in the process of generating a closure. 

Is this something to worry about now? Could a bug be easily introduced? Is there ever going to be a case where we are generating a closure type inside a generic type?

TIHan replied:
If we can figure out how to make this correct, that would be better and prevent any further issues from cropping up.

dsyme replied:
The 0 here represents that a closure doesn't reside within a generic class - there are no "enclosing class type parameters" to lop off.  

I've removed the TODO and added that comment

TIHan replied:
Great, that makes a lot of sense.

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -4860,10 +5054,25 @@ and GenDelegateExpr cenv cgbuf eenvouter expr (TObjExprMethod((TSlotSig(_, deleg
 
 /// Generate statically-resolved conditionals used for type-directed optimizations.
 and GenStaticOptimization cenv cgbuf eenv (constraints, e2, e3, _m) sequel =
+    // Note: during IlxGen, even if answer is StaticOptimizationAnswer.Unknown we discard the static optimization
```
Great explanation!

TIHan commented on src/fsharp/IlxGen.fs:
```diff
@@ -5892,10 +6136,13 @@ and GenMethodForBinding
-                if HasFSharpAttribute g g.attrib_NoDynamicInvocationAttribute v.Attribs then
-                    let exnArg = mkString g m (FSComp.SR.ilDynamicInvocationNotSupported(v.CompiledName g.CompilerGlobalState))
+                let attr = TryFindFSharpBoolAttributeAssumeFalse cenv.g cenv.g.attrib_NoDynamicInvocationAttribute v.Attribs
+                if (not generateWitnessArgs && attr.IsSome) ||
```
nit: Pattern match be better?
```fsharp
match attr, generateWitnessArgs with
| Some _, false
| Some false, true ->
    ...
| _ ->
    ...
```

dsyme replied:
In this case I prefer the if/then/else I think

TIHan commented on src/fsharp/MethodOverrides.fs:
```diff
@@ -153,7 +153,7 @@ module DispatchSlotChecking =
     let GetObjectExprOverrideInfo g amap (implty, id: Ident, memberFlags, ty, arityInfo, bindingAttribs, rhsExpr) = 
         // Dissect the type
-        let tps, argInfos, retTy, _ = GetMemberTypeInMemberForm g memberFlags arityInfo ty id.idRange
+        let tps, _, argInfos, retTy, _ = GetMemberTypeInMemberForm g memberFlags arityInfo 0 ty id.idRange
```
The "0" reminds me of the "TODO: 0 may be wrong here" comment since it is representing the number of enclosing typars. As far as I know this should be ok? I know I'm being redundant in asking, but I just want to make sure. :)

dsyme replied:
I double checked, this one is OK

TIHan commented on src/fsharp/QuotationTranslator.fs:
```diff
@@ -620,12 +712,46 @@ and private ConvExprCore cenv (env : QuotationTranslationEnv) (expr: Expr) : QP.
+                if g.generateWitnesses && inWitnessPassingScope then 
+                    match env.witnessesInScope.TryGetValue traitInfo.TraitKey with 
+                    | true, storage -> Some storage
+                    | _ -> None // failwithf "no storage for witness %s found in scope" w.MemberName
```
nit: Remove commented code.

If this actually ever returned None, should it actually throw though?

dsyme replied:
Removed the commented code.  The None case does assert below - I think I'd prefer to code this defensively as is (because so many important scenarios still continue to work even if witnesses generation returns None due to a witness not being found correctly in the environment)

TIHan commented on src/fsharp/TcGlobals.fs:
```diff
@@ -1497,6 +1509,84 @@ type public TcGlobals(compilingFslib: bool, ilg:ILGlobals, fslibCcu: CcuThunk, d
 
+  /// Find an FSharp.Core LaguagePrimitives dynamic function that corresponds to a trait witness, e.g.
+  /// AdditionDynamic for op_Addition.  Also work out the type instantiation of the dynamic function.
+  member __.makeBuiltInWitnessInfo (t: TraitConstraintInfo) =
```
nit: Just a question, while other members in TcGlobals have Pascal-cased names, this one and the one below are camel-cased. What is the reasoning behind that? Just curious.

dsyme replied:
Yes it's inconsistent, there were other functions on TcGlobals that used that casing.

TIHan commented on src/fsharp/TypedTreeOps.fs:
```diff
@@ -2259,8 +2283,35 @@ let checkMemberVal membInfo arity m =
+
+let GetTopValTypeInCompiledForm g topValInfo numEnclosingTypars ty m =
     let tps, paramArgInfos, rty, retInfo = GetTopValTypeInFSharpForm g topValInfo ty m
+    let witnessInfos = GetTraitWitnessInfosOfTypars g numEnclosingTypars tps // TODO: parentTypars
```
This has a "TODO", is if fair to expand on this a bit further? Will doing this "TODO" be for new language features in the future and/or break compat?

dsyme replied:
I'll take a look

dsyme replied:
I removed the TODO - it was fixed in an earlier iteration

dsyme commented:
> I think once the nits and questions are resolved I will mark this as approved and can be merged. :)

resolved! :)

TIHan commented:
@dsyme @cartermp There a few more nits + questions that I had.

dsyme commented:
I fixed the build break here

dsyme commented:
@TIHan My apologies for missing so many of your comments!  I've addressed them all now

cartermp commented:
From March 20th 2019 until June 6th 2020 - hooray!

dsyme commented:
@alfonsogarciacaro asked this:

> @dsyme I was hoping that this or #6810 would extend the TypedTree with information that could help Fable (or other compilers) resolve the SRTPs, but I see BasicPatterns.TraitCall remains unchanged so it's going to be very difficult to support this feature.

I will look at this now.

This is actually also a really motiviatio for getting this issue addressed - it allows FCS-based compilers like Fable to be written that are accurate w.r.t. the intended semantics of F#
